packages/http-server/src/request-handlers/submit-order.ts

```diff
@@ -27,14 +27,34 @@ export function submitOrder(opts: SubmitOrderOpts): RequestHandler {
       return res.status(400).json({ errors: [errorResponse] })
     }
 
-    // TODO: return 409 if order is not allowed given the current state of the exchange. (#issue 4)
+    const exchange = await exchangesApi.getExchange({id: message.exchangeId})
+    if(exchange == undefined) {
+      const errorResponse: ErrorDetail = { detail: `exchangeId: ${message.exchangeId} is undefined` }
 
-    const quote = await exchangesApi.getQuote({ exchangeId: message.exchangeId })
+      return res.status(404).json({ errors: [errorResponse] })
+    }
+
+    console.log('hiiiii')
+
+    const last = exchange[exchange.length-1]
+    if(!last.validNext.has('order')) {
+      const errorResponse: ErrorDetail = { detail: `cannot submit Order for an exchange where the last message is kind: ${last.kind}` }
+
+      return res.status(409).json({ errors: [errorResponse] })
+    }
+
+    const quote = exchange.find((message) => message.isQuote) as Quote
     if(quote == undefined) {
       const errorResponse: ErrorDetail = { detail: 'quote is undefined' }
       return res.status(404).json({errors: [errorResponse]})
     }
 
+    if(new Date(quote.expiresAt) < new Date(message.createdAt)){
```

Comment:
nit: I guess technically, the message creator could spoof the `createdAt` timestamp to get around this check

```suggestion
    if(new Date(quote.expiresAt) < new Date()){
```